File: 프로그래머스/lv2/12924. 숫자의 표현/숫자의 표현.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int n) {
    int answer = 0;
    
    vector<int> numbers;
    
    for (int i = 1; i <= n; i++) {
        numbers.push_back(i);
    }
    
    int left = 0;
    int right = 0;
    int acc = numbers[left];
    
    while (left <= right && right < n) {
        if (left != right) {
            if (acc < n) {  // 작을 때
                right++;
                if (right < n) {
                    acc += numbers[right];
                }
            }
            else if (acc > n) { // 클 때
                acc -= numbers[left];
                left++;
            }
            else {
                answer++;
                right++;
                if (right < n) {
                    acc += numbers[right];
                }
            }
        }
        else {
            if (acc == n) {
                answer++;
            }
            right++;
            if (right < n) {
                acc += numbers[right];
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/2/12924. 숫자의 표현/숫자의 표현.java
import java.util.*;

class Solution {
    int[] numbers;
    
    public int solution(int n) {
        int answer = 0;
        
        numbers = new int[n];
        
        for (int i = 0; i < n; i++) {
            numbers[i] = i + 1;
        }
        
        answer = solve(n);
        
        return answer;
    }
    
    int solve(final int number) {
        int result = 0;
        int left = 0;
        int right = 0;
        int accumulator = numbers[left];
        
        while (left <= right && inRange(right)) {
            if (left == right) {
                if (number == accumulator) {
                    result++;
                }
                
                right++;
                if (inRange(right)) {
                    accumulator += numbers[right];
                }
            }
            else {
                if (number == accumulator) {
                    result++;
                    
                    right++;
                    if (inRange(right)) {
                        accumulator += numbers[right];
                    }
                }
                else if (number > accumulator) {
                    right++;
                    if (inRange(right)) {
                        accumulator += numbers[right];
                    }
                }
                else {
                    accumulator -= numbers[left];
                    left++;
                }
            }
        }
        
        return result;
    }
    
    boolean inRange(final int index) {
        if (0 <= index && index < numbers.length) {
            return true;
        }
        
        return false;
    }
}